### PR TITLE
feat(enhancement): New Effects for Shield Damage, Overheating, and Cloaking

### PR DESCRIPTION
**Pull request by Koranir**
**Feature:** This PR implements the feature request detailed and discussed in issue #7069

## Feature Details
This PR adds a modified effect based on the cloak effect that makes ships glow blue when their shields are damaged, and red when they overheat. The cloak effect has also been changed to a semi-transparent grey with red outlines.

The shield glow and overheat are toggleable with the Damage highlights setting.

## UI Screenshots
image

Cloak:

https://user-images.githubusercontent.com/101683475/199621890-d3fb65fe-40bd-4195-8378-068893672d1f.mp4

Shield:

https://user-images.githubusercontent.com/101683475/199621911-b9f45cd2-b79c-4dae-8126-f9f6dc9e692b.mp4

Heat:

https://user-images.githubusercontent.com/101683475/199417764-b7423953-6d36-46a4-9ad4-d9c6ab78238d.mp4

## Testing Done
I flew around and shot some ships with flamethrowers, lasers and particle cannons, and there were no visual glitches that I could find.
Multiple species and swizzles were tested, and they all work well. 

## Performance Impact
There is a moderate performance hit, as each ship in active combat essentially gets rendered 2-3 more times, but for non-potatoes it is fine.

### CODE REVIEW

**Comment by Koranir** [comment]
#2568 as well, kind of.

**Comment by Zireael07** [comment]
Looks awesome plus you know you did hit the ship <3

**Comment by EjoThims** [comment]
Toning down the brightness and making shield effect look better would be far better, but I think it'd be best even then to have that toggle defaulted to off.

**Comment by Zitchas** [comment]
I like the looks of this. II think the colors might be a bit bright, so toning it down a little would be good. But yeah, I like it. The color should lightly tint the ship, not overwhelm it. If a ship gets that red, it's pretty much slag and everyone inside cooked...

**Comment by Hurleveur** [comment]
Agreed, that much red should be for when a ship is twice over its heat limit.

**Comment by Saugia** [comment]
There's quite a lot happening all at once in that footage- could you please have separate examples of each effect? It would help make it easier to review each color and have less distracting things going on otherwise.

**Review by GefullteTaubenbrust2 — CHANGES_REQUESTED** [review]
Tbh not insanely thrilled about it. I think it is a really cool idea, but the additive doesn’t look that great in my opinion, at least for the shield effect. The one for heat looks good, although I agree it should only show for severe overheating. 
But for the shield one, it just feels kind of slapped on. Part of the problem probably is the color, which i feel should be a brighter blue. But it doesn’t really occur to me why damaging shields would make the ship glow blue. I‘d really like to see more fidelity in this effect, like a sort of fresnel effect at least, although obviously that would probably be really difficult to implement. 
And given the performance impact, I‘m really not sure how I feel about it. So idk if anything, have it for the heat damage, although even then, how realistic is it you could heat a ship to the point of glowing red hot without completely destroying it? At least for the shields, I’d just stick with the normal effects system, if even that.

**Comment by GefullteTaubenbrust2** [comment]
Maybe if it was more subtle I’d be more forgiving. Part of the reason I find it to look so odd is just how strong it is. But again, I am somewhat concerned about the performance impact.

**Comment by GefullteTaubenbrust2** [comment]
Maybe if it was more subtle I’d be more forgiving. Part of the reason I find it to look so odd is just how strong it is. The performance impact could be resolved by making it optional, so we could address that too.

**Comment by Saugia** [comment]
It'd definitely have to be something, if implemented, disabled by default.

**Comment by Koranir** [comment]
> It'd definitely have to be something, if implemented, disabled by default.

Yep, it's set to off by default

**Comment by RisingLeaf** [comment]
What I think would be kind of nice, is to apply some sort of noise to both of these to make them look less status and more , not realistic, but interesting I guess

**Review by Saugia — CHANGES_REQUESTED** [review]
Overall thoughts:

I have to agree with Gefu, I'm not terribly invested in these effects. Endless Sky is a game that incorporates an arcade-like design while maintaining some semi-realistic aspects, but I feel these effects are a bit *too* arcade-like. For example, there was some talk a long time ago about using "status markers" for effects instead of visuals, such as "down arrows" for slow damage, etc., but was ultimately not something we want to represent the visuals of damage with, leaning more towards individual effects. Other games do occasionally rely on this sort of thing (large, glowey shields), but in ES's environment it feels a bit to visually messy and can get overwhelming. 

However, there is a history for requests for this sort of thing, so it's not that it's unwanted. Hence why I suggested it being default to disabled, and also more importantly, disabled for users that it would increase the load for. Also, it wouldn't be the first interface usage that is somewhat screen-hogging (particularly the shield circles etc., I don't play with these). So I can see this being implemented as a togglable option. In such a case:

The Cloak looks pretty good, I like that it makes the ship slightly transparent now and the grey is a lot less obnoxious than a deep dark red, and pops out a bit more due to the lightness. It could be just the tiniest bit darker though, especially due to the fact most ships in the game use light greys (Coalition is a good example). Some players may get confused whether or not their ship is cloaked, especially zoomed out, but the subtle transparency might alleviate that. Though, the odd thing to me is this almost feels like something we'd do to the cloak for the base toggles because it adds depth, and then shift the single-color non-transparent glow it currently has to this toggle because it's more simple.

The heating one stands out a lot more unfortunately. I have to agree with the other comments saying it seems way too vibrant, and doesn't really make sense mechanically + visually as a ship that is that bright red and hot would, at that point, be effectively non-functional and dead [fried crew ):]. But being this is a concept for more simplistic visuals to represent damage types and visual clarification, it doesn't necessarily need to be removed, but the red should be toned *way* down.

The shield one is a bit of a perplexing one. By making an effect for this, you are double-downing on the lore that ships' shields are wrapped around the hull. However, you are also scaling them up slightly, which doubles the amount of outline every ship has and that can once again clutter the screen. Additionally, by making them blue, you are now suggesting lore-wise that all shield colors are this teal color. Do you think that make sense? Personally, for example, I would think other factions like the Korath and Remnant have shields in different colors. Once again this comes down to whether or not we want to use this as purely a game tool as blue is typically the generic sci-fi color of shields, or have it be actually part of the ships in-game (regardless of toggle). Agree on other comments though- the teal seems odd. Make it more of a light blue. Right now it seems a bit too hard to see for an effect designed to show visual clarification.

A side note- be sure you test these effects with already existing damage-type effects in the game such as corrosion (parts of the hull breaking off), heat damage (hot smoke clouds coming off the ship), disruption (white sparking effect), etc.. I'm most concerned that these effects, combined with the ones in this PR, will get too visually chaotic and defeat the purpose.

**Comment by MasterOfGrey** [comment]
This really feels like an accessibility feature, like colourblind friendly interfaces and other things like that.

As a feature I don't mind it, but it should come with a tooltip warning or something about the performance hit if it's turned on in the options.
And I think if we implemented it it would be time to separate the Display section into two parts: 

- one for actual display preferences (zoom, screen mode, vsync, planet labels, mini-map, rotating flagship
- an Accessibility section for status overlays, player flagship highlighting, and these shields and heat options (which I think should be separated)

Also, as an aside, the "flash" part of the cloak effect is not one that I like as a default experience. I know that wasn't you though.

**Comment by Koranir** [comment]
>Personally, for example, I would think other factions like the Korath and Remnant have shields in different colors.

So would I, but as far as I know there is no way to get the original faction of a ship, so unless somebody goes and adds a "faction" or "shieldType" attribute to the every ship in the game, one colour is all I can code for.

**Review by Saugia — CHANGES_REQUESTED** [review]
Trailing off my and others' review comments, I'll now treat these purely as visual indicators and not part of in-game design/lore; more of a display/interface option. 

Cloak: Much better. More visual clarification.

Heat: Much better; a lot less intense. Regardless of analysis based on lore (bright red = very hot for a ship), this works nice as a visual indicator now and isn't too crazy. Though, it seems just the *tiniest* bit too little now, if you feel like bumping it up just a few units would help.

Shield: Still feeling a little iffy about this one. Since the heat one was brought down to look nicer, the shield one does look better with comparison now, but I feel the same applies- it is slightly too weak of an effect still. I'm not quite sure if just a little less transparency, or just a little more brightness, is the way to go. Play around with it a little (and maybe match it to the heat one).

It may be an accident, but testing the heat effect with a pirate helps because pirates have red in their swizzle. To match that, it might benefit you to to test the shield effect on merchant ships. Though, now that I think about it, Pug ships would also be very useful to test the two effects on with each swizzle as their entire ships are colored.

**Comment by Zireael07** [comment]
Love the cloak and heat. Shield looks a little bit too weak, seconding the previous poster.

File: source/Engine.cpp [hunk 1]
```diff
@@ -951,6 +951,15 @@ void Engine::Draw() const
...
 	}
 
+	for(const shared_ptr<Ship> &ship : ships)
+	{
+		if(ship->Cloaking() && ship->IsYours() && ship->GetSystem() == player.GetSystem())
+		{
+			OutlineShader::Draw(ship->GetSprite(), (ship->Position()-center-ship->Velocity())*zoom, Point(ship->Width(), ship->Height())*zoom,
+								Color(0.6f, 0.1f, 0.1f, 0.8f*static_cast<float>(ship->Cloaking())), ship->Facing().Unit(), ship->GetFrame());
+		}
+	}
```
Comment (Hurleveur) [inline source/Engine.cpp:961]: You dont need any of these braces since its one line code within another one line code

File: source/Engine.cpp [hunk 2]
```diff
@@ -2322,15 +2331,31 @@ void Engine::AddSprites(const Ship &ship)
 {
 	bool hasFighters = ship.PositionFighters();
 	double cloak = ship.Cloaking();
+	bool damageHighlight = Preferences::Has("Damage highlights");
 	bool drawCloaked = (cloak && ship.IsYours());
+	bool drawHeat = ship.Heat() > 0.9;
+	bool drawShield = ship.RecentShield() > 4.;
+	double shield = sqrt(ship.RecentShield()/ship.Attributes().Get("shields"));
+	double heat = min((ship.Heat()-0.9) / 16, 1.);
```
Comment (Hurleveur) [inline source/Engine.cpp:2339]:
```suggestion
	double heat = min((ship.Heat()  - .9) / 16, 1.);
```
Just a bit of spacing
Btw 0.9 could be written .9 since its the same.
Also I think it should be put in the nameless namespace at the start of engine as a constant (HIGH_HEAT or smth I'm not good with names)

File: source/Engine.cpp [hunk 3]
```diff
@@ -2322,15 +2331,31 @@ void Engine::AddSprites(const Ship &ship)
...
-	auto drawObject = [&itemsToDraw, cloak, drawCloaked](const Body &body) -> void
+	auto drawObject = [&itemsToDraw, cloak, shield, heat, drawCloaked, drawShield, drawHeat, damageHighlight](const Body &body) -> void
 	{
-		// Draw cloaked/cloaking sprites swizzled red, and overlay this solid
-		// sprite with an increasingly transparent "regular" sprite.
+		// Draw cloaked/cloaking sprites swizzled transparent grey.
 		if(drawCloaked)
-			itemsToDraw.AddSwizzled(body, 27);
-		itemsToDraw.Add(body, cloak);
+			itemsToDraw.AddSwizzled(body, 10, 0.7);
```
Comment (Hurleveur) [inline source/Engine.cpp:2345]: Not sure whats going on with the drawing changes and if they're necessary since I dont know that much about that so would be grateful it smbd that does could have a look.
You do realize drawCloaked was to show only player ships that are cloaked in red, and not enemy ships?

Whatever the case you need to adjust the indentation

File: source/Ship.cpp [hunk 4]
```diff
@@ -2212,6 +2212,10 @@ void Ship::DoGeneration()
 	energy -= ionization;
 	fuel -= leakage;
 	heat += burning;
+
+	recentShield *= 0.75;
+	recentHeat *= 0.984;
```
Comment (Hurleveur) [inline source/Ship.cpp:2217]: Those numbers seem to pop out of nowhere... They should be constant in the nameless namespace at the start of the file. Also I'm wondering why we bother to calculate recent heat at all; shoudnt we just show how heated the ship is currently?

Also really not a fan of doing these calculations when the preference is disabled, please check for that.

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Conceptually I like the idea, however there are a few things that either need clarification or be changed

Visually I quite like the new cloak and overheat, the shields look ok as well, but thats not my area

File: source/Engine.cpp (hunk 3 above)
Comment (Saugia) [reply source/Engine.cpp:2345]: Probably better we don't have enemy ships outlined while cloaking, that's not so much visual clarification as it is almost cheating by outlining them more. The idea is to just do it for player ships.

File: source/Ship.cpp (hunk 4 above)
Comment (Koranir) [reply source/Ship.cpp]: It might be more costly to check the preferences every time for one operation?

File: source/Engine.cpp [hunk 5]
```diff
@@ -951,6 +955,25 @@ void Engine::Draw() const
 			RingShader::Draw(pos, radius, 1.5f, it.disabled, color[6 + it.type], dashes, it.angle);
 	}
 
+	Point dCenter = player.FlagshipPtr()->Position();
+	for(const shared_ptr<Ship> &ship : ships)
```
Comment (quyykk) [inline source/Engine.cpp:959]: Each each ship to draw is drawn in `AddSprites`. Why not move the loop body there, since that's where each ship is drawn. No need to loop through every ship again.

**Review by Saugia — COMMENTED** [review]
Looking good now visually. I'm going to defer the rest to a code/content dev.

File: source/SpriteShader.cpp [hunk 6]
```diff
@@ -74,8 +74,9 @@ namespace {
 		{GL_GREEN, GL_BLUE, GL_GREEN, GL_ALPHA}, // 24 darkened magenta
 		{GL_RED, GL_BLUE, GL_RED, GL_ALPHA}, // 25 pure magenta
 		{GL_RED, GL_GREEN, GL_RED, GL_ALPHA}, // 26 faded magenta
-		{GL_BLUE, GL_ZERO, GL_ZERO, GL_ALPHA}, // 27 red only (cloaked)
```
Comment (quyykk) [inline source/SpriteShader.cpp:77]: The red only swizzle is still useful, no need to remove it imo. (also backwards compatibility)

File: source/SpriteShader.cpp [hunk 7]
```diff
@@ -74,8 +74,9 @@ namespace {
...
-		{GL_ZERO, GL_ZERO, GL_ZERO, GL_ALPHA} // 28 black only (outline)
+		{GL_ALPHA, GL_GREEN, GL_BLUE, GL_ALPHA}, // 27 bright red (heat)
```
Comment (quyykk) [inline source/SpriteShader.cpp:77]: here it's AGBA, but in the fallback swizzle it's ABBA?

File: source/SpriteShader.cpp [hunk 8]
```diff
@@ -244,11 +245,14 @@ void SpriteShader::Init(bool useShaderSwizzle)
...
 		"      break;\n"
 		"    case 27:\n"
-		"      color = vec4(color.b, 0.f, 0.f, color.a);\n"
+		"      color = color.abba;\n"
 		"      break;\n"
 		"    case 28:\n"
 		"      color = vec4(0.f, 0.f, 0.f, color.a);\n"
 		"      break;\n"
+		"    case 29:\n"
+		"      color = vec4(color.r*0.43f, color.r*0.55f, color.r*0.70, color.a);\n"
```
Comment (quyykk) [inline source/SpriteShader.cpp:254]: Here the fallback swizzle is different than the hardware one.

File: source/Ship.h [hunk 9]
```diff
@@ -577,6 +578,8 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
 	// Delays for shield generation and hull repair.
 	int shieldDelay = 0;
 	int hullDelay = 0;
+	// Recent damage taken.
+	double recentShield = 0.;
```
Comment (quyykk) [inline source/Ship.h:582]: `recentShieldDamage` is more clear imo

Comment (Hurleveur) [reply source/Ship.h:582]: True, often variables can be made explicit instead of requiring a comment and it's always smth that should be done. Idk how I missed it

File: source/Engine.cpp (hunk 5 above)
Comment (Koranir) [reply source/Engine.cpp]: Because the effects makes use of the outline shader, it gets drawn directly when it's called. `AddSprites`, however, places all the sprites into a vector that is later drawn in the `Engine::Draw()` function. `Engine::Draw()` is also where the outline shader is right now, after all the sprites in the vector are drawn. If the outline shader goes into `AddSprites()`, it will later be drawn over by the ship's sprite, and have no effect.

TL;DR:
The outline shader only works after the rest of the sprites are drawn under it, which is in the`Draw()` function. Moving it to `AddSprites` will break it, as `AddSprites` is not where the sprites are drawn.

File: source/Engine.cpp [hunk 10]
```diff
@@ -928,6 +932,25 @@ void Engine::Draw() const
 	draw[drawTickTock].Draw();
 	batchDraw[drawTickTock].Draw();
 
+	Point dCenter = player.FlagshipPtr()->Position();
+	for(const shared_ptr<Ship> &ship : ships)
+	{
+		if(ship->Cloaking() && ship->IsYours() && ship->GetSystem() == player.GetSystem())
+		{
+			float cloak = .1f * static_cast<float>(ship->Cloaking());
+			OutlineShader::Draw(ship->GetSprite(), (ship->Position()-dCenter)*zoom, Point(ship->Width(), ship->Height())*zoom,
```
Comment (Hurleveur) [inline source/Engine.cpp:941]:
```suggestion
			OutlineShader::Draw(ship->GetSprite(), (ship->Position() - dCenter) * zoom, Point(ship->Width(), ship->Height()) * zoom,
```

File: source/Engine.cpp [hunk 11]
```diff
@@ -928,6 +932,25 @@ void Engine::Draw() const
...
+								Color(6 * cloak, cloak, cloak, 8 * cloak), ship->Facing().Unit(), ship->GetFrame());
+		}
+
+		if(Preferences::Has("Damage highlights"))
+		{
+			float shields = min(SHIELD_OUTLINE_MULT*static_cast<float>(sqrt(ship->RecentShieldDamage()/ship->Attributes().Get("shields"))), 1.f);
+			if(ship->RecentShieldDamage() > 4. && ship->GetSystem() == player.GetSystem())
+				OutlineShader::Draw(ship->GetSprite(), (ship->Position()-dCenter)*zoom, Point(ship->Width(), ship->Height())*zoom,
```
Comment (Hurleveur) [inline source/Engine.cpp:949]:
```suggestion
				OutlineShader::Draw(ship->GetSprite(), (ship->Position() - dCenter) * zoom, Point(ship->Width(), ship->Height()) * zoom,
```

File: source/Engine.cpp [hunk 12]
```diff
@@ -928,6 +932,25 @@ void Engine::Draw() const
...
+									Color(.61f * shields, .78f * shields, shields, shields),ship->Facing().Unit(), ship->GetFrame());
```
Comment (Hurleveur) [inline source/Engine.cpp:950]:
```suggestion
									Color(.61f * shields, .78f * shields, shields, shields), ship->Facing().Unit(), ship->GetFrame());
```

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Just some spaces

**Review by beccabunny — APPROVED** [review]
Great visuals, i like it

**Comment by Zireael07** [comment]
Shield looks great now <3

File: source/Engine.cpp [hunk 13]
```diff
@@ -2322,15 +2345,31 @@ void Engine::AddSprites(const Ship &ship)
...
+			if(drawHeat)
+				itemsToDraw.AddSwizzled(body, 30);
+			// Overlay this solid sprite with an increasingly transparent "regular" sprite.
+			itemsToDraw.Add(body, max(cloak, drawHeat ? heat : 0.));
+			// Draw another the sprite scaled up swizzled blue over when the shields are damaged.
+			if(drawShield)
+				itemsToDraw.AddSwizzled(body, 29, max(1-shield, 0.75));
+		}
+		else // We would have missed on this step if damageHighlight was disabled.
+			itemsToDraw.Add(body, cloak);
```
Comment (Amazinite) [inline source/Engine.cpp:2372]: We generally avoid inline comments.
```suggestion
		else
		{
			// We would have missed on this step if damageHighlight was disabled.
			itemsToDraw.Add(body, cloak);
		}
```
This would technically work without the braces, but it looks spooky to me that way.

File: source/Engine.cpp [hunk 14]
```diff
@@ -2322,15 +2345,31 @@ void Engine::AddSprites(const Ship &ship)
 {
 	bool hasFighters = ship.PositionFighters();
 	double cloak = ship.Cloaking();
+	bool damageHighlight = Preferences::Has("Damage highlights");
 	bool drawCloaked = (cloak && ship.IsYours());
+	bool drawHeat = ship.Heat() > 0.9;
+	bool drawShield = ship.RecentShieldDamage() > 4.;
+	double shield = sqrt(ship.RecentShieldDamage()/ship.Attributes().Get("shields"));
```
Comment (Amazinite) [inline source/Engine.cpp:2352]:
```suggestion
	double shield = sqrt(ship.RecentShieldDamage() / ship.Attributes().Get("shields"));
```

File: source/Engine.cpp [hunk 15]
```diff
@@ -928,6 +932,25 @@ void Engine::Draw() const
...
+		if(ship->Cloaking() && ship->IsYours() && ship->GetSystem() == player.GetSystem())
+		{
+			float cloak = .1f * static_cast<float>(ship->Cloaking());
+			OutlineShader::Draw(ship->GetSprite(), (ship->Position() - dCenter) * zoom, Point(ship->Width(), ship->Height()) * zoom,
+								Color(6 * cloak, cloak, cloak, 8 * cloak), ship->Facing().Unit(), ship->GetFrame());
+		}
+
+		if(Preferences::Has("Damage highlights"))
+		{
+			float shields = min(SHIELD_OUTLINE_MULT*static_cast<float>(sqrt(ship->RecentShieldDamage()/ship->Attributes().Get("shields"))), 1.f);
```
Comment (Amazinite) [inline source/Engine.cpp:947]:
```suggestion
			float shields = min(SHIELD_OUTLINE_MULT * static_cast<float>(sqrt(ship->RecentShieldDamage() / ship->Attributes().Get("shields"))), 1.f);
```

File: source/Engine.cpp [hunk 16]
```diff
@@ -928,6 +932,25 @@ void Engine::Draw() const
...
+	for(const shared_ptr<Ship> &ship : ships)
+	{
+		if(ship->Cloaking() && ship->IsYours() && ship->GetSystem() == player.GetSystem())
+		{
+			float cloak = .1f * static_cast<float>(ship->Cloaking());
+			OutlineShader::Draw(ship->GetSprite(), (ship->Position() - dCenter) * zoom, Point(ship->Width(), ship->Height()) * zoom,
+								Color(6 * cloak, cloak, cloak, 8 * cloak), ship->Facing().Unit(), ship->GetFrame());
+		}
+
+		if(Preferences::Has("Damage highlights"))
```
Comment (Hurleveur) [inline source/Engine.cpp:945]: How about we store this in a bool before the loop instead of checking for each ship?

File: source/Engine.cpp [hunk 17]
```diff
@@ -928,6 +932,25 @@ void Engine::Draw() const
...
 	batchDraw[drawTickTock].Draw();
 
+	Point dCenter = player.FlagshipPtr()->Position();
+	for(const shared_ptr<Ship> &ship : ships)
+	{
+		if(ship->Cloaking() && ship->IsYours() && ship->GetSystem() == player.GetSystem())
+		{
+			float cloak = .1f * static_cast<float>(ship->Cloaking());
+			OutlineShader::Draw(ship->GetSprite(), (ship->Position() - dCenter) * zoom, Point(ship->Width(), ship->Height()) * zoom,
+								Color(6 * cloak, cloak, cloak, 8 * cloak), ship->Facing().Unit(), ship->GetFrame());
```
Comment (warp-core) [inline source/Engine.cpp:942]:
```suggestion
								Color(6.f * cloak, cloak, cloak, 8.f * cloak), ship->Facing().Unit(), ship->GetFrame());
```

File: source/Engine.cpp [hunk 18]
```diff
@@ -928,6 +932,25 @@ void Engine::Draw() const
 	draw[drawTickTock].Draw();
 	batchDraw[drawTickTock].Draw();
 
+	Point dCenter = player.FlagshipPtr()->Position();
```
Comment (warp-core) [inline source/Engine.cpp:935]: This is a nice way to get a segfault when the player has no flagship.

File: source/Engine.cpp [hunk 19]
```diff
@@ -928,6 +932,25 @@ void Engine::Draw() const
...
+		}
+
+		if(Preferences::Has("Damage highlights"))
+		{
+			float shields = min(SHIELD_OUTLINE_MULT * static_cast<float>(sqrt(ship->RecentShieldDamage() / ship->Attributes().Get("shields"))), 1.f);
+			if(ship->RecentShieldDamage() > 4. && ship->GetSystem() == player.GetSystem())
+				OutlineShader::Draw(ship->GetSprite(), (ship->Position() - dCenter) * zoom, Point(ship->Width(), ship->Height()) * zoom,
+									Color(.61f * shields, .78f * shields, shields, shields), ship->Facing().Unit(), ship->GetFrame());
+		}
+	}
```
Comment (warp-core) [inline source/Engine.cpp:952]: Put the two if statements inside the for loop into their own lambdas.
Then, do:
```c++
if(Preferences::Has("Damage highlights")
    for(<every ship>)
        if(<ship is in player system>)
        {
            CloakLambda(ship);
            ShieldDamageHighlight(ship);
        }
else
    for(<every ship>)
        if(<ship is in player system>)
            CloakLambda(ship);
```

File: source/Engine.cpp [hunk 20]
```diff
@@ -2322,15 +2345,34 @@ void Engine::AddSprites(const Ship &ship)
 {
 	bool hasFighters = ship.PositionFighters();
 	double cloak = ship.Cloaking();
+	bool damageHighlight = Preferences::Has("Damage highlights");
 	bool drawCloaked = (cloak && ship.IsYours());
+	bool drawHeat = ship.Heat() > 0.9;
+	bool drawShield = ship.RecentShieldDamage() > 4.;
+	double shield = sqrt(ship.RecentShieldDamage() / ship.Attributes().Get("shields"));
```
Comment (warp-core) [inline source/Engine.cpp:2352]: It might be cheaper to add a field to `Ship` that can be used to store this, and set this value there (via a public setter method) instead of recalculating it later in `Engine::Draw()`.

File: source/Engine.cpp [hunk 21]
```diff
@@ -2322,15 +2345,34 @@ void Engine::AddSprites(const Ship &ship)
...
+		if(damageHighlight)
+		{
+			// Draw the sprite swizzled red when overheating.
+			if(drawHeat)
+				itemsToDraw.AddSwizzled(body, 30);
+			// Overlay this solid sprite with an increasingly transparent "regular" sprite.
+			itemsToDraw.Add(body, max(cloak, drawHeat ? heat : 0.));
+			// Draw another the sprite scaled up swizzled blue over when the shields are damaged.
+			if(drawShield)
+				itemsToDraw.AddSwizzled(body, 29, max(1-shield, 0.75));
```
Comment (warp-core) [inline source/Engine.cpp:2369]:
```suggestion
				itemsToDraw.AddSwizzled(body, 29, max(1. - shield, 0.75));
```

**Review by warp-core — CHANGES_REQUESTED** [review]

**Comment by Koranir** [comment]
Superceded by #8400
